### PR TITLE
feat(framework) Implement `DriverServicer.Pull/PushServerAppInputs/Outputs`

### PR DESCRIPTION
**Pull request by panh99**
Based on:

1. #4363
2. #4365

### CODE REVIEW

File: src/py/flwr/server/superlink/driver/driver_servicer.py [hunk 1]
```diff
@@ -208,13 +204,47 @@ def PullServerAppInputs(
...
+                return PullServerAppInputsResponse()
+
+            serverapp_ctxt = state.get_serverapp_context(run_id)
+            run = state.get_run(run_id)
+            fab = None
+            if run and run.fab_hash:
+                if result := ffs.get(run.fab_hash):
+                    fab = Fab(run.fab_hash, result[0])
+
+            if serverapp_ctxt and run and fab:
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/driver/driver_servicer.py:226]: what if `serverapp_ctxt` is `None`?

File: src/py/flwr/server/superlink/driver/driver_servicer.py [hunk 2]
```diff
@@ -208,13 +204,47 @@ def PullServerAppInputs(
         self, request: PullServerAppInputsRequest, context: grpc.ServicerContext
     ) -> PullServerAppInputsResponse:
         """Pull ServerApp process inputs."""
-        raise NotImplementedError()
+        log(DEBUG, "DriverServicer.PullServerAppInputs")
+        state = self.state_factory.state()
+        ffs = self.ffs_factory.ffs()
+
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/driver/driver_servicer.py:210]: let's add a single line comment of what lines below do?

File: src/py/flwr/server/superlink/driver/driver_servicer.py [hunk 3]
```diff
@@ -208,13 +204,47 @@ def PullServerAppInputs(
...
+        with self.lock:
+            if request.HasField("run_id"):
+                run_id: Optional[int] = request.run_id
+            else:
+                run_id = state.get_pending_run_id()
+            if run_id is None:
+                return PullServerAppInputsResponse()
+
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/driver/driver_servicer.py:218]: let's add a single line comment of what lines below do?

**Review by jafermarq — APPROVED** [review]

File: src/py/flwr/server/superlink/driver/driver_servicer.py (hunk 2 above)
Comment (panh99) [reply src/py/flwr/server/superlink/driver/driver_servicer.py:210]: Sure!

File: src/py/flwr/server/superlink/driver/driver_servicer.py (hunk 3 above)
Comment (panh99) [reply src/py/flwr/server/superlink/driver/driver_servicer.py:218]: Good idea!

File: src/py/flwr/server/superlink/driver/driver_servicer.py (hunk 1 above)
Comment (panh99) [reply src/py/flwr/server/superlink/driver/driver_servicer.py:226]: oops, my bad. The ServerApp context can be `None` here, and we will just return `None` to the `flwr-serverapp` process.

Comment (panh99) [reply src/py/flwr/server/superlink/driver/driver_servicer.py]: I've updated the code.

**Review by danieljanes — APPROVED** [review]

**Pull request merged**